### PR TITLE
support lock list

### PR DESCRIPTION
**Pull request by liqinggd**

### CODE REVIEW

File: rcore-fs/src/vfs.rs [hunk 1]
```diff
@@ -386,6 +396,24 @@ pub trait FileSystem: Sync + Send {
...
+/// INodeLockListCreater is used to create INodeLockList for Filesystems
+pub trait INodeLockListCreater: Send + Sync {
+    fn new_empty_list(&self) -> Arc<dyn INodeLockList>;
+}
+
+/// INodeLockList is used to manage some inode locks.
+/// User can implement own specific type of INodeLockList and inode lock,
+/// then use `downcast_ref` to cast the INodeLockList to the specific type.
+pub trait INodeLockList: Any + Send + Sync {
+    fn as_any_ref(&self) -> &dyn Any;
```
Comment (tatetian) [inline rcore-fs/src/vfs.rs]: This method is redundant given the fact that that `InodeLockList` implements the `Any` trait, the latter of which provides the `downcast_ref` method.

File: rcore-fs-sefs/src/lib.rs [hunk 2]
```diff
@@ -695,6 +717,7 @@ impl SEFS {
         device: Box<dyn Storage>,
         time_provider: &'static dyn TimeProvider,
         uuid_provider: &'static dyn UuidProvider,
+        lock_list_creater: Option<Box<dyn INodeLockListCreater>>,
```
Comment (tatetian) [inline rcore-fs-sefs/src/lib.rs]: Do you think it make sense to take an argument of `&'static dyn INodeLockListCreater` here to save one heap allocation?

File: rcore-fs-sefs/src/lib.rs [hunk 3]
```diff
@@ -622,6 +627,21 @@ impl vfs::INode for INodeImpl {
...
+            let new_lock_list = self.fs.alloc_lock_list()?;
+            *lock_list = Some(new_lock_list);
+        }
+        Ok(lock_list.as_ref().unwrap().clone())
+    }
+
+    fn test_lock_list(&self) -> Option<Arc<dyn INodeLockList>> {
+        let lock_list = self.lock_list.read();
+        lock_list.as_ref().and_then(|l| Some(l.clone()))
+    }
```
Comment (tatetian) [inline rcore-fs-sefs/src/lib.rs:806]: Is it possible to combine the two methods into one? I know the intention of adding the `test_lock_list` method in addition to the `lock_list` method is to do lazy initialization of `InodeLockList`. But this complicates the interface. The interface of inodes and files are already very complex. I think it is highly desirable to encapsulate all lock list-related logic (here, the lazy initialization) inside `InodeLockList` itself.

File: rcore-fs/src/vfs.rs [hunk 4]
```diff
@@ -95,6 +95,16 @@ pub trait INode: Any + Sync + Send {
...
 
+    /// Get the lock list of the INode, create an empty one if not exist
+    fn lock_list(&self) -> Result<Arc<dyn INodeLockList>> {
+        Err(FsError::NotSupported)
+    }
+
+    /// Test if INode has lock list
+    fn test_lock_list(&self) -> Option<Arc<dyn INodeLockList>> {
+        None
+    }
```
Comment (tatetian) [inline rcore-fs/src/vfs.rs:106]: I am not a big fan of the proposed changes to the `Inode` trait. Conceptually, lock lists are notions of OSes, not FSes. You also acknowledge this fact. That is why you have made an admirable effort to introduce the `INodeLockList` trait to decouple the implementations of lock lists (which is part of the OSes) from the interface (which is part of the VFS). But the result is not perfect in my opinion. The VFS layer stills needs to know the concept of `INodeLockList`, which is a trivial trait that virtually says nothing about lock lists, except that they are `Any`. Can we do better?

I think I have an alternative solution, which can completely remove the notion of lock lists from VFS. The solution is based on `Extensions`, which is a generalized and also extensible version of the current solution.

```rust
pub trait Inode {
    /// Get the extension of this inode.
    fn ext(&self) -> &Extension;
}

/// An extension is a set of objects that is attached to
/// an inode.
///
/// Each objects of an extension is of different types.
/// In other words, types are used as the keys to get and
/// set the objects in an extension.
//
// For implementation details, read [this blog post](https://www.jakobmeier.ch/blogging/Untapped-Rust.html)
// and see the [AnyMap](https://crates.io/crates/anymap) crate.
pub struct Extension(/* ... */);

impl Extension {
    pub fn new() -> Self {
        todo!("implement new")
    }

    /// Get an object of `Arc<T>`.
    pub fn get<T: Any + Send + Sync>(&self) -> Option<Arc<T>> {
        todo!("implement get")
    }

    /// Put an object of `Arc<T>`. If there exists one object of the type,
    /// then the old one is returned.
    pub fn put<T: Any + Send + Sync>(&self, obj: Arc<T>) -> Option<Arc<T>> {
        todo!("implement set")
    }

    /// Delete an object of `Arc<T>`. If there exists one object of the type,
    /// then the old one is returned.
    pub fn del<T: Any + Send + Sync>(&self) -> Option<Arc<T>> {
        todo!("implement del")
    }
}
```

**Comment by liqinggd** [comment]
Please refer to https://github.com/occlum/sefs/pull/29

**Pull request closed without merging**